x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/mapper/ReservedRoleMappingXContentNameFieldHelper.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.authc.support.mapper;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Helper class to avoid name erasure when cluster-state role mappings on disk.
+ * When cluster-state role mappings are written to disk, their XContent format is used.
+ * This format omits the role mapping's name.
+ * This means that if CS role mappings are ever recovered from disk (e.g., during a master-node restart), their names are erased.
+ * To address this, this class augments CS role mapping serialization to persist the name of a mapping in a reserved metadata field,
+ * and recover it from metadata during de-serialization.
+ * Storing the name in metadata allows us to persist the name without BWC-breaks in role mapping `XContent` format (as opposed to changing
+ * XContent serialization to persist the top-level name field).
+ * It also ensures that role mappings are re-written to cluster state in the new,
+ * name-preserving format the first time operator file settings are processed.
+ * {@link #copyWithNameInMetadata(ExpressionRoleMapping)}} is used to copy the name of a role mapping into its metadata field.
+ * This is called when the role mappings are read from operator file settings (at which point the correct name is still available).
+ * {@link #parseWithNameFromMetadata(XContentParser)} is used to parse a role mapping from XContent restoring the name from metadata.
+ */
```

Comment:
I like that everything related to this fix is in one class, but I think it becomes a little hard to follow what's going on and it would be nice to not have to do a helper class if it's not needed. A short comment for each of the methods and putting them in the class where they're used would be better. 

I also think the constants belong in `RoleMappingMetadata` since that class is already responsible for representing a role mapping as xcontent.

Reply:
Fair! I'll move the code into `RoleMappingMetadata`

Reply:
I _do_ want to keep `copyWithNameInMetadata` and `parseWithNameFromMetadata` in the same class since they really only make sense together. Let me know if you have strong objections.